src/common/wrap/wrap-websocket.js

```diff
@@ -3,59 +3,167 @@
+        ? function (...args) {
+          try {
+            return listener.apply(this, args)
+          } catch (error) {
+            wsInstance.#tagError(error)
+            throw error
+          }
+        }
+        : listener?.handleEvent
+          ? { // case for listener === object with handleEvent
+              handleEvent: function (...args) {
+                try {
+                  return listener.handleEvent.apply(listener, args)
+                } catch (error) {
+                  wsInstance.#tagError(error)
+                  throw error
+                }
+              }
+            }
+          : listener // case for listener === null
+      return super.addEventListener(type, wrappedListener, ...rest)
     }
 
-    send (...args) {
-      this.report('send', ...args)
+    send (data) {
+      // Only track metrics if the connection is OPEN; data sent in CONNECTING state throws, and data sent in CLOSING/CLOSED states is silently discarded
+      if (this.readyState === WebSocket.OPEN) {
+        const { type, size } = getDataInfo(data)
+        this.nrData.sendCount = (this.nrData.sendCount ?? 0) + 1
+        this.nrData.sendBytes = (this.nrData.sendBytes ?? 0) + size
+        this.nrData.sendBytesMin = Math.min(this.nrData.sendBytesMin ?? Infinity, size)
+        this.nrData.sendBytesMax = Math.max(this.nrData.sendBytesMax ?? 0, size)
+        if (!(this.nrData.sendTypes ?? '').includes(type)) {
+          this.nrData.sendTypes = this.nrData.sendTypes ? `${this.nrData.sendTypes},${type}` : type
+        }
+      }
       try {
-        return super.send(...args)
-      } catch (err) {
-        this.report('send-err', ...args)
-        throw err
+        return super.send(data)
+      } catch (error) {
+        this.#tagError(error)
+        throw error
+      }
+    }
+
+    close (...args) {
+      try {
+        super.close(...args)
+      } catch (error) {
+        this.#tagError(error)
+        throw error
       }
     }
   }
 
   globalScope.WebSocket = WrappedWebSocket
-  return sharedEE
+  return wsEE
+}
+
+/**
+ * Returns the data type and size of the WebSocket send data
+ * @param {*} data - The data being sent
+ * @returns {{ type: string, size: number }} - The type name and size in bytes
+ */
+function getDataInfo (data) {
+  if (typeof data === 'string') {
+    return {
+      type: 'string',
+      size: new TextEncoder().encode(data).length // efficient way to calculate the # of UTF-8 bytes that WS sends (cannot use string length)
+    }
+  }
+  if (data instanceof ArrayBuffer) {
+    return { type: 'ArrayBuffer', size: data.byteLength }
+  }
+  if (data instanceof Blob) {
+    return { type: 'Blob', size: data.size }
+  }
+  if (data instanceof DataView) {
+    return { type: 'DataView', size: data.byteLength }
+  }
+  if (ArrayBuffer.isView(data)) {
+    return { type: 'TypedArray', size: data.byteLength }
+  }
+  return { type: 'unknown', size: 0 }
```

Comment:
We should discuss fail path here, ie what to do with the socket instance if we cant calculate the message size since that is a major utility driver on this feature.

Reply:
Lets just plan to revisit this if it becomes an issue later